pkg/apis/core/validation/validation.go

```diff
@@ -1945,14 +1957,41 @@ func validateObjectFieldSelector(fs *core.ObjectFieldSelector, expressions *sets
 
 	if len(fs.APIVersion) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath.Child("apiVersion"), ""))
-	} else if len(fs.FieldPath) == 0 {
+		return allErrs
+	}
+	if len(fs.FieldPath) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath.Child("fieldPath"), ""))
-	} else {
-		internalFieldPath, _, err := legacyscheme.Scheme.ConvertFieldLabel(fs.APIVersion, "Pod", fs.FieldPath, "")
-		if err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("fieldPath"), fs.FieldPath, fmt.Sprintf("error converting fieldPath: %v", err)))
-		} else if !expressions.Has(internalFieldPath) {
-			allErrs = append(allErrs, field.NotSupported(fldPath.Child("fieldPath"), internalFieldPath, expressions.List()))
+		return allErrs
+	}
+
+	internalFieldPath, _, err := podshelper.ConvertDownwardAPIFieldLabel(fs.APIVersion, fs.FieldPath, "")
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("fieldPath"), fs.FieldPath, fmt.Sprintf("error converting fieldPath: %v", err)))
+		return allErrs
+	}
+
+	path, subscript := fieldpath.SplitMaybeSubscriptedPath(internalFieldPath)
+	if len(subscript) > 0 {
+		// This is to indicate that the internalFieldPath has a subscript, so
+		// that we can compare the path against the allowed path set easily.
+		path += "[]"
```

Comment:
triggering off of the length of the subscript means that `metadata.annotations['']` and `metadata.annotations` are indistinguishable

trying to fit dynamic expressions into a fixed list of valid inputs isn't a good fit, and results in difficult to understand errors as noted in https://github.com/kubernetes/kubernetes/pull/55902/files#r152679054

I'd expect completely separate paths for subscripted and fixed expressions:

```go
if path,subscript,ok := fieldpath.SplitSubscriptedPath(internalFieldPath); ok {
  switch path {
    case "metadata.annotations":
      ...
    case "metadata.labels":
      ...
    default:
      ... path does not support subscript error
  }
} else if !expressions.Has(internalFieldPath) {
  allErrs = append(allErrs, field.NotSupported(fldPath.Child("fieldPath"), path, expressions.List()))
  return allErrs
}
```

Reply:
Done.